Feature/rails3 add cukes

ekabadayi opened the pull request:
added some basic cukes, methods and paths for menu navigation.

ekabadayi committed:
added: basic cukes for user activation, authorization, registration and user settings. modified: user_factory, menu_steps, general_steps and custom_web_steps.

ekabadayi committed:
added: newlines at at the end of the code

tessi commented:
Scenario: A user is unable to register if one of the constraints left blank (features/users/user_registration.feature:24) seems to fail.

ekabadayi committed:
resetted to a stable state. error message for user_registration modified

ekabadayi committed:
Merge branch 'feature/rails3_add_cukes' of github.com:ekabadayi/openproject into feature/rails3_add_cukes

Conflicts:
	features/support/paths.rb

ekabadayi committed:
basic project access cuke added

ekabadayi committed:
small update on register and activation cukes

ekabadayi committed:
password length has been modified

ekabadayi committed:
initial cuke for groups added

ekabadayi committed:
Merge branch 'feature/rails3' of https://github.com/opf/openproject into feature/rails3_add_cukes

ekabadayi committed:
project accession cuke according visibility settings added.

simple group deletion cuke added.  some feature names has been modified

tessi commented on features/groups/group_deletion.feature:
```diff
@@ -0,0 +1,18 @@
+Feature: Simple deletion of a group
+
+Background:
```
I think it would be better to add a 'Given we have the group "Bob's Team"' step. You can use at the beginning of the Scenario. The current Background tests group creation (which is already done by the group_initial.feature) - and clicking through the UI makes this test much slower.

tessi commented on features/projects/project_visibility.feature:
```diff
@@ -0,0 +1,30 @@
+Feature: Project Visibility
+
+Background:
+	Given I am logged in as "admin"
```
Instead of using this background step, you could use the following step definitions defined in general_steps.rb:

<pre><code>Given /^there is 1 [pP]roject with(?: the following)?:$/ do |table|
  p = FactoryGirl.build(:project)
  send_table_to_object(p, table)
end

Then /^the project "([^"]*)" is( not)? public$/ do |project_name, negation|
  p = Project.find_by_name(project_name)
  p.update_attribute(:is_public, !negation)
end</code></pre>

tessi commented on features/step_definitions/custom_web_steps.rb:
```diff
@@ -51,3 +51,10 @@
   page.execute_script "jQuery(#{selector.inspect}).removeClass('hover');"
 end
+
+When /^(?:|I )continue with "([^"]*)" in "([^"]*)"$/ do |link, css_class|
```
isn't the "in" part in this step definition covered by the "within (.*)" step?

tessi commented on features/step_definitions/general_steps.rb:
```diff
@@ -68,6 +68,13 @@
 end
 
+When(/^I log me in as "(.*?)" with password "(.*?)"$/) do |username, password|
+  FactoryGirl.create :admin unless User.where(:login => 'admin').any?
```
the password param is not forwarded to <code>login()</code> - why not?

If you intended to not forward it, why not use the "Given /^(?:|I )am logged in as "([^\"]*)"$/ do |username|" step defined in general_steps.rb ?

tessi commented on features/step_definitions/general_steps.rb:
```diff
@@ -68,6 +68,13 @@
+  FactoryGirl.create :admin unless User.where(:login => 'admin').any?
+  FactoryGirl.create :anonymous unless AnonymousUser.count > 0
+  #provokes a unsuccessful login by using any username but admin
+  login(username, 'admin')
```
the default admin pw is now "adminADMIN!" (we extended the minimal password length and complexity)

tessi commented on features/step_definitions/menu_steps.rb:
```diff
@@ -17,6 +17,16 @@
   nodes.first.click if nodes.present?
 end
 
+# opens a drop-down menu.
```
should be something like "#opens a menu item in the main menu", as it is not a drop-down :)

EDIT: Right now I see that there are multiple occurrences of the "menu_root" css class. One in the project menu, and another in the drop-down menu. Not sure which is the "first" (as you do a nodes.first.click). You might consider using a xpatch query to better identify the menu_root item you want.

tessi commented on features/users/user_auth.feature:
```diff
@@ -0,0 +1,27 @@
+Scenario: A user is able to login successfully with provided credentials
+Given I am on the login page
+And I am not logged in
+And I am admin
```
No need for the "And I am not logged in"-step when you do a "And I am admin", or am I wrong?

tessi commented on features/users/user_auth.feature:
```diff
@@ -0,0 +1,27 @@
+And I am not logged in
+And I am admin
+And I open the "Openproject Admin" menu
+Then I should see "My account" within ".menu_root" 
```
".menu_root" is ambiguous - we have it in the main menu and in the project menu

tessi commented on features/users/user_project_access.feature:
```diff
@@ -0,0 +1,15 @@
+And I am logged in as "tom"
+And the user "tom" is a "super_user" in the project "P1"
+And I open the "Projects" menu
+And  I follow "View all projects" within ".menu_root"
```
".menu_root" is ambiguous - we have it in the main menu and in the project menu

tessi commented on features/users/user_registration.feature:
```diff
@@ -0,0 +1,44 @@
+Background: 
+Given I am on the homepage
+And I open the "Sign in" menu
+And I follow "Register" within ".menu_root"
```
".menu_root" is ambiguous - we have it in the main menu and in the project menu

tessi commented on features/users/user_settings.feature:
```diff
@@ -0,0 +1,21 @@
+Scenario: A user is able to change his name if the settings permit it
+Given I am admin
+And   I open the "Openproject Admin" menu
+And   I follow "My account" within ".menu_root"
```
".menu_root" is ambiguous - we have it in the main menu and in the project menu

tessi commented on features/users/user_settings.feature:
```diff
@@ -0,0 +1,21 @@
+Scenario: A user is able to change his mail address if the settings permit it
+Given I am admin
+And   I open the "Openproject Admin" menu
+Then  I should see "My account" within ".menu_root"
```
".menu_root" is ambiguous - we have it in the main menu and in the project menu

tessi commented on features/users/user_settings.feature:
```diff
@@ -0,0 +1,21 @@
+Given I am admin
+And   I open the "Openproject Admin" menu
+Then  I should see "My account" within ".menu_root"
+And   I follow "My account" within ".menu_root"
```
".menu_root" is ambiguous - we have it in the main menu and in the project menu

tessi commented:
Hi Erdem, Thanks for your Pull Request! :)

can you please merge the current feature/rails3 into your PR? I see that you test some login stuff, which we changed lately. Just want to make sure the tests pass with a current feature/rails3

ekabadayi committed:
Merge branch 'feature/rails3' of https://github.com/opf/openproject into feature/rails3_add_cukes

ekabadayi committed:
update according to review for feature/rails3 add cukes (#154)

unnecessary steps has been removed, several changes has been made
in order to improve the test efficiency

ekabadayi commented:
Hi Tessi, thank you for your Review!

I have updated my PR and tests seem all passing in current feature/rails3

tessi commented:
erdem, could you be so kind to rebase your PR on feature/rails3 again? we cannot merge atm, as there are conflicts.

mfrister commented:
@ekabadayi: Any more interest in fixing this pull request? No updates for three months.

ekabadayi commented:
Thanks for the reminder. I have revised this PR in #689. Hope that works.

ulferts commented:
I am closing this PR because of the new PR (#689).

ulferts commented:
@ekabadayi, @tessi: please always consider not using selenium (@javascript) as the first option. Otherwise our cukes will be slowed down unnecessarily.
